pkg/controller/nodes/executor.go

```diff
@@ -659,11 +676,32 @@ func (c *nodeExecutor) RecursiveNodeHandler(ctx context.Context, execContext exe
 
 		// This is an optimization to avoid creating the nodeContext object in case the node has already been looked at.
 		// If the overhead was zero, we would just do the isDirtyCheck after the nodeContext is created
-		nodeStatus := nl.GetNodeExecutionStatus(ctx, currentNode.GetID())
 		if nodeStatus.IsDirty() {
 			return executors.NodeStatusRunning, nil
 		}
 
+		// Now if the node is of type task, then let us check if we are within the parallelism limit, only if the node
+		// has been queued already
+		if currentNode.GetKind() == v1alpha1.NodeKindTask && nodeStatus.GetPhase() == v1alpha1.NodePhaseQueued {
+			maxParallelism := execContext.GetExecutionConfig().MaxParallelism
+			if maxParallelism > 0 {
+				// If we are queued, let us see if we can proceed within the node parallelism bounds
+				if execContext.CurrentParallelism() >= maxParallelism {
+					logger.Infof(ctx, "Maximum Parallelism for task nodes achieved [%d] >= Max [%d], Round will be short-circuited.", execContext.CurrentParallelism(), maxParallelism)
+					return executors.NodeStatusRunning, nil
```

Comment:
why do you proceed to running if the current parallelism **exceeds** the max?

Reply:
we dont proceed to running, this is the nodeStatus :(, this implies the node is running so check it again. But the task - start is skipped

Reply:
ah okay sorry i got confused by what proceed meant in the comment above